index/test/rtree/rtree_with_strategies.cpp

```diff
@@ -0,0 +1,156 @@
+                            || bg::util::is_point<QueryValue>::value
+                                && bg::util::is_point<typename Rtree::value_type>::value;
+
+template
+<
+    typename QueryValue, typename Rtree,
+    std::enable_if_t<is_bp_or_bb_v<QueryValue, Rtree>, int> = 0
+>
+void test_queries_bp_bb(Rtree const& rtree)
+{
+    std::vector<typename Rtree::value_type> values;
+    rtree.query(bgi::covered_by(QueryValue{}), std::back_inserter(values));
+    rtree.query(bgi::disjoint(QueryValue{}), std::back_inserter(values));
+    rtree.query(bgi::within(QueryValue{}), std::back_inserter(values));
+}
+
+template
+<
+    typename QueryValue, typename Rtree,
+    std::enable_if_t<(!is_bp_or_bb_v<QueryValue, Rtree>), int> = 0
+>
+void test_queries_bp_bb(Rtree const& ) {}
+
+template
+<
+    typename QueryValue, typename Rtree,
+    std::enable_if_t<is_pb_or_bb_v<QueryValue, Rtree>, int> = 0
+>
+void test_queries_pb_bb(Rtree const& rtree)
+{
+    // These predicates use algorithms that are not implemented for
+    // some geometry combinations
+    std::vector<typename Rtree::value_type> values;
+    rtree.query(bgi::contains(QueryValue{}), std::back_inserter(values));
+    rtree.query(bgi::covers(QueryValue{}), std::back_inserter(values));
+}
+
+template
+<
+    typename QueryValue, typename Rtree,
+    std::enable_if_t<(!is_pb_or_bb_v<QueryValue, Rtree>), int> = 0
+>
+void test_queries_pb_bb(Rtree const& ) {}
+
+template
+<
+    typename QueryValue, typename Rtree,
+    std::enable_if_t<is_pp_or_bb_v<QueryValue, Rtree>, int> = 0
+>
+void test_overlaps_pp_bb(Rtree const& rtree)
+{
+    std::vector<typename Rtree::value_type> values;
+    rtree.query(bgi::overlaps(QueryValue{}), std::back_inserter(values));
+    rtree.query(bgi::touches(QueryValue{}), std::back_inserter(values));
+}
+
+template
+<
+    typename QueryValue, typename Rtree,
+    std::enable_if_t<(!is_pp_or_bb_v<QueryValue, Rtree>), int> = 0
+>
+void test_overlaps_pp_bb(Rtree const& ) {}
+
+template <typename QueryValue, typename Rtree>
+void test_queries(Rtree const& rtree)
+{
+    std::vector<typename Rtree::value_type> values;
+    rtree.query(bgi::intersects(QueryValue{}), std::back_inserter(values));
+    rtree.query(bgi::nearest(QueryValue{}, 1), std::back_inserter(values));
+    test_queries_bp_bb<QueryValue>(rtree);
+    test_queries_pb_bb<QueryValue>(rtree);
+    test_overlaps_pp_bb<QueryValue>(rtree);
+}
+
+template <typename Value, typename Params, typename Strategies>
+void test_strategies()
+{
+    std::vector<Value> values;
+    bgi::rtree<Value, bgi::parameters<Params, Strategies>> rtree{values};
+    rtree.insert(Value{});
+    rtree.remove(Value{});
+
+    test_queries<point>(rtree);
+    test_queries<box>(rtree);
+    test_queries<segment>(rtree);
+}
+
+template <typename Value, typename Params>
+void test_params()
+{
+    test_strategies<Value, Params, bg::strategies::index::cartesian<>>();
+    test_strategies<Value, Params, bg::strategies::cartesian<>>();
```

Comment:
What is the difference actually? Is it another umbrella?

Reply:
Yes, just like various algorithms has their own strategies index has it as well. And in addition to that it should work with the highest level one which can be passed anywhere.